### PR TITLE
monty_lab migration

### PR DESCRIPTION
**Pull request by scottcanoe**
This PR brings in everything from the dmc directory from monty_lab. The only changes are a couples of lines to the README (which we'll want to extend when we have more publicly available data, etc.)

I've rebuilt the environment using instructions in `monty/README.md` and verified that experiments run. A future PR will address any problems, if they come up, while rerunning all experiments and reanalyzing the data.

### CODE REVIEW

File: monty/configs/__init__.py [hunk 1]
```diff
@@ -1,8 +1,20 @@
-# Copyright 2025 Thousand Brains Project
-#
-# Copyright may exist in Contributors' modifications
-# and/or contributions to the work.
-#
-# Use of this source code is governed by the MIT
-# license that can be found in the LICENSE file or at
-# https://opensource.org/licenses/MIT.
```
Comment (tristanls) [inline monty/configs/__init__.py:8]: **issue**: We still need the copyright.

Comment (scottcanoe) [reply monty/configs/__init__.py]: Got it, thanks.

File: README.md [hunk 2]
```diff
@@ -6,4 +6,28 @@ This repository contains code to replicate experiments from our paper, "Thousand
...
+
+```
+DMC_ROOT_DIR/
+    pretrained_models/
+    results/
+    view_finder_images/
+    visualizations/
+```
+
+The `pretrained_models` directory contains the pre-trained models for Monty experiments. The `results` directory contains the results of evaluation experiments. The `view_finder_images` directory contains the images used for input to ViT-based models as well as for some figure visualizations.
```
Comment (vkakerbeck) [inline README.md:23]: Will we upload these files onto AWS storage for people to download, similar to the monty meets world dataset and benchmark pretrained models?

File: monty/configs/README.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+## Figure 1 & 2: Diagramatic Figures With No Experiments
+
+## Figure 3: Robust Sensorimotor Inference
+
+This figure presents results from four inference experiments testing Monty's robustness under different conditions. Monty was pre-trained on 14 standard rotations derived from cube face and corner views (see full configuration details in `pretraining_experiments/dist_agent_1lm`).
+
+- `dist_agent_1lm`: Standard inference with no sensor noise or random rotations
+- `dist_agent_1lm_noise`: Tests robustness to sensor noise
+- `dist_agent_1lm_randrot_all`: Tests performance across 14 random rotations
```
Comment (vkakerbeck) [inline monty/configs/README.md:13]:
```suggestion
- `dist_agent_1lm_randrot_all`: Tests performance across 14 random rotations, not seen during training
```

File: monty/configs/README.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+- NO hypothesis-testing*
+- No voting
+- Varying numbers of rotations trained on (evaluations use different baseline models)
+
+*No hypothesis-testing as the ViT model comparison only receives one view and cannot move around object, and no noise since Sensor-Module noise does not have a clear analogue for the ViT model.
+
+The main output measure is accuracy and rotation error as a function of training rotations.
+
+**Notes:**
+- Training rotations should be structured:
```
Comment (vkakerbeck) [inline monty/configs/README.md:118]:
```suggestion
- Training rotations are ordered as:
```

File: README.md [hunk 5]
```diff
@@ -6,4 +6,28 @@ This repository contains code to replicate experiments from our paper, "Thousand
...
+
+Figures and tables are stored under `DMC_ANALYSIS_DIR`. By default, this is `~/tbp/results/dmc_analysis`, but you can change this by setting the `DMC_ANALYSIS_DIR` environment variable.
+
+## Running the Experiments
+
+Configs to train models and run experiments are in the `configs` directory.
```
Comment (vkakerbeck) [inline README.md:29]: It could be good to include brief instructions on how to run these configs. Just a basic command or at least a link to the Monty docs on that.

**Review by vkakerbeck — APPROVED** [review]
Wow, super impressive! Really clean and nicely structured 🚀
I just left a few minor comments on the readme but nothing that should prevent merging this.

File: README.md [hunk 6]
```diff
@@ -6,4 +6,28 @@ This repository contains code to replicate experiments from our paper, "Thousand
 
 Experiments make use of either the Monty framework, or Pytorch in the case of deep learning models. Instructions for environment setup and experiment execution can be found in the README files within `monty` and `pytorch`.
```
Comment (tristanls) [inline README.md:7]: **issue (non-blocking)**: This sets the expectation of a `monty/README.md`, but such a file is not present.

Comment (scottcanoe) [reply README.md:7]: Do you mean `pytorch/README.md`? `monty/README.md` exists but may not show up in default review view since its unchanged (along with `environment.yml`, etc.)

File: README.md (hunk 2 above)
Comment (scottcanoe) [reply README.md:23]: Yeah, I think that's the plan for pretrained models. I hadn't thought about putting results and ViT input somewhere, but it makes sense and would be easy to do. I'm re-running experiments right now from this repo, so after that I think @hlee9212 and I can coordinate on moving pretty everything into AWS storage that anyone would want or need to run experiments or just analysis scripts.

File: monty/configs/view_finder_images.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+
+"""
+Configs
+---------
+"""
+
+# Define the 14 standard training rotations used for 'view_finder_base'.
+train_rotations = get_cube_face_and_corner_views_rotations()
+
+# Define out motor system config that uses the custom policy and uses
```
Comment (nielsleadholm) [inline monty/configs/view_finder_images.py:304]:
```suggestion
# Define our motor system config that uses the custom policy and uses
```

File: monty/configs/view_finder_images.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+# a shorter desired object distance than default.
+motor_system_config = MotorSystemConfigInformedNoTransStepS20(
+    motor_system_class=FramedObjectPolicy,
+    motor_system_args=make_informed_policy_config(
+        action_space_type="distant_agent_no_translation",
+        action_sampler_class=ConstantSampler,
+        rotation_degrees=5.0,
+        use_goal_state_driven_actions=False,
+        switch_frequency=1.0,
+        good_view_percentage=0.5,  # Make sure we define the required
```
Comment (nielsleadholm) [inline monty/configs/view_finder_images.py:314]: `# Make sure we define the required` - not sure if this needs updating or just removing?

File: scripts/data_utils.py [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+        translation = np.asarray(translation)
+        out = deepcopy(self)
+        out.pos += translation
+        return out
+
+    def __sub__(self, translation: ArrayLike) -> "ObjectModel":
+        translation = np.asarray(translation)
+        return self + (-translation)
+
+
```
Comment (nielsleadholm) [inline scripts/data_utils.py:479]: **suggestion**: Can remove some of this white space.

File: scripts/fig2.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        ax = fig.add_subplot(1, 1, 1)
+        ax.imshow(image)
+        ax.axis("off")
+        fig.tight_layout(pad=0)
+        fig.savefig(png_dir / f"{i}.png")
+        fig.savefig(svg_dir / f"{i}.svg", bbox_inches="tight", pad_inches=0)
+        plt.show()
+
+
+def plot_dist_agent_pretraining_epochs():
```
Comment (nielsleadholm) [inline scripts/fig2.py:121]: We can technically remove these. We will have them in git history if we ever want to return to plotting something like this.

File: scripts/fig4.py [hunk 11]
```diff
@@ -0,0 +1,849 @@
...
+        rotations = load_symmetry_rotations(stats)
+        for r in rotations + [target, rand]:
+            r.pose_error = np.degrees(
+                get_pose_error(r.rot.as_quat(), target.rot.as_quat())
+            )
+
+        # - Find mlh, best, and some other symmetric.
+        rotations = sorted(rotations, key=lambda x: x.pose_error)
+        min_ = sorted(rotations, key=lambda x: x.pose_error)[0]
+        sym = rotations[np.random.randint(1, len(rotations))]
```
Comment (nielsleadholm) [inline scripts/fig4.py:225]: **question/suggestion:** Is a seed set anywhere? Would be worth doing otherwise this result will vary with each run.

**Review by nielsleadholm — APPROVED** [review]
Thanks for putting this together, looks great! Tried to take this as an opportunity to have another general look through the code, so just highlighted a few things we can clean up.

File: scripts/fig2.py (hunk 10 above)
Comment (scottcanoe) [reply scripts/fig2.py:121]: These `plot_*_pretraining_epochs` functions are used to generate the actual plots in figure 2 though (where models are shown after x number of epochs). Is the argument for deletion because they're not strictly data analysis?

File: monty/configs/README.md [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+Here we are showing the performance of the "standard" version of Monty, using:
+- 77 objects
+- 14 rotations
+- Goal-state-driven/hypothesis-testing policy active
+- A single LM (no voting)
+
+The main output measures are accuracy, rotation error (degrees), and Chamfer distance for each condition.
+
+Figure 3 also include an analysis of symmetric rotations, the data for which is generated with the experiment `fig3_symmetry_run` which is defined in `visualizations.py`.
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply monty/configs/README.md:24]: Right. So rather than adding a ton of comments about visualization experiments, I'm thinking it'd be cleanest to focus this part of README primarily on main configs. There's a Visualization Experiments section which is pretty comprehensive, and the files in the `scripts` directory have complete descriptions of which experiments are needed for which panels. Does that sound good? (i.e., removing text like this in the main README body)

File: monty/configs/README.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+  2. Next 8 rotations = cube corners
+  3. Remaining = random rotations (as otherwise introduces redundancy)
+
+## Figure 8: Computationally Efficient Learning and Inference
+
+### Inference (12 experiments)
+
+There are two sets of experiments, one using hypothesis testing and another using no hypothesis testing.
+
+- `dist_agent_1lm_randrot_nohyp_x_percent_5p` - 5% threshold (No Hypothesis Testing)
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply monty/configs/README.md]: Good catch, I simplified it to what you wrote. I think we need the second config though since I believe this is the only experiment with 5 random rotations and no noise. I'll double check though.

File: README.md (hunk 5 above)
Comment (scottcanoe) [reply README.md:29]: Good idea, I added text in the top-level section.

File: monty/configs/README.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+- Training rotations are ordered as:
+  1. First 6 rotations = cube faces
+  2. Next 8 rotations = cube corners
+  3. Remaining = random rotations (as otherwise introduces redundancy)
+
+## Figure 8: Computationally Efficient Learning and Inference
+
+### Inference (12 experiments)
+
+There are two sets of experiments, one using hypothesis testing and another using no hypothesis testing.
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply monty/configs/README.md]: I committed this change manually.

File: monty/configs/fig8_flops.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+# Copyright may exist in Contributors' modifications
+# and/or contributions to the work.
+#
+# Use of this source code is governed by the MIT
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+
+"""Configs for Figure 8: Flops Comparison.
+
+This module defines the following experiments:
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply monty/configs/fig8_flops.py:12]: Yeah, I've tried not to touch anything @hlee9212 is working on to avoid conflicts. I think it makes sense that we leave a follow up PR to @hlee using these comments as a reference for safety if that's cool with everyone.

File: monty/configs/README.md [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+All of these experiments should be run in serial due to the memory needs of
+detailed logging (or checkpoint-saving in the case of
+`fig2_pretrain_surf_agent_1lm_checkpoints`).
+
+All experiments save their results to subdirectories of `DMC_ROOT` / `visualizations`.
+
+## Other Experiments
+`view_finder_experiments.py` defines five experiments:
+- view_finder_base: 14 standard training rotations
+- view_finder_randrot_all: 14 randomly generated rotations
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply monty/configs/README.md:204]: I've checked with @hlee9212, and it looks like this is indeed unused. Deleted.

File: scripts/render_view_finder_images.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+Monty. To run this script, call
+
+```
+$ python render_view_finder_images.py -e EXPERIMENT_NAME
+```
+
+Running this script creates the directory
+`EXPERIMENT_DIR/view_finder_rgbd/visualizations` where `EXPERIMENT_DIR` is one of
+ - `view_finder_base`: 14 standard training rotations
+ - `view_finder_randrot_all`: 14 randomly generated rotations
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply scripts/render_view_finder_images.py:25]: Yep! Done.

File: monty/configs/view_finder_images.py (hunk 7 above)
Comment (scottcanoe) [reply monty/configs/view_finder_images.py:303]: Fixed manually.

File: monty/configs/view_finder_images.py (hunk 8 above)
Comment (scottcanoe) [reply monty/configs/view_finder_images.py:313]: Right, it wasn't actually the relevant parameter (desired object distance is, which is described in the comment above. I've deleted this line comment.

File: scripts/fig4.py (hunk 11 above)
Comment (scottcanoe) [reply scripts/fig4.py:225]: Good point. I've set the seed at the top of this file.

File: README.md (hunk 6 above)
Comment (tristanls) [reply README.md:7]: I meant `monty/README.md`, but as you point out. It is there. Not sure what I was looking at. 👍

File: monty/configs/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (nielsleadholm) [reply monty/configs/README.md]: Yeah good idea, sounds good!

File: scripts/fig2.py (hunk 10 above)
Comment (nielsleadholm) [reply scripts/fig2.py:121]: Ah my bad, ignore me.

**Review by nielsleadholm — APPROVED** [review]
Perfect, thanks for making those changes.
